src/Controller/CRUDController.php

```diff
@@ -78,12 +183,12 @@ public function setContainer(?ContainerInterface $container = null): void
      *
      * @param string               $view       The view name
      * @param array<string, mixed> $parameters An array of parameters to pass to the view
-     *
-     * @return Response A Response instance
      */
-    public function renderWithExtraParams($view, array $parameters = [], ?Response $response = null)
+    public function renderWithExtraParams($view, array $parameters = [], ?Response $response = null): Response
     {
-        return $this->render($view, $this->addRenderExtraParams($parameters), $response);
+        $response = $response ?? new Response();
+
+        return $response->setContent($this->twig->render($view, $this->addRenderExtraParams($parameters)));
```

Comment:
Since `Response` was added, it would be nice to add `string` to `$view` and remove it from phpdoc.

What about change the variable name from `$view` to `$template`? Looks better to me.

Reply:
IMO `string` should not be added. Better will be add `TemplateWrapper` to phpdoc. Look at:
https://github.com/twigphp/Twig/blob/3.x/src/Environment.php#L266-L278

Reply:
> IMO `string` should not be added. Better will be add `TemplateWrapper` to phpdoc. Look at:
> https://github.com/twigphp/Twig/blob/3.x/src/Environment.php#L266-L278

It's true about the `Environment::render`, but I think it's kind of weird to have an instance of `TemplateWrapper` in the controller.

Anyway, why not using the `AbstractController::render` method?

Reply:
> Anyway, why not using the `AbstractController::render` method?

https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Controller/CRUDController.php#L53-L56

IMO We better will be remove inheritance from `AbstractController` and split Actions to seperate files under `src/Action/CRUD` directory(simmilar to CoreController).

Reply:
> IMO better will be remove inheritance from `AbstractController` and split Actions to seperate files under `src/Action/CRUD` directory(simmilar to CoreController).

Sure, I agree, but AFAIK that is not happening in 4.0. 

> https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Controller/CRUDController.php#L53-L56

Since now we are extending `AbstractController` we have access to `$this->render` method, [in Symfony 5.1 directly in the class](https://github.com/symfony/symfony/blob/5.1/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php#L255) and [in Symfony 4.4 through the ControllerTrait](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php#L39).

Reply:
> > IMO better will be remove inheritance from `AbstractController` and split Actions to seperate files under `src/Action/CRUD` directory(simmilar to CoreController).
> 
> Sure, I agree, but AFAIK that is not happening in 4.0.

What about:
- based on this in `3.x` we can add actions (like CoreController in 3.x)
- in `4.x` CRUDController 
    - inject actions instead service 
    - remove inheritance from `AbstractController`
    - add CRUDControllerInterface
    - make CRUDController final

After this change people will be able to easy override CRUDController in thier admins. (or part of it).

@sonata-project/contributors WDYT? It is simply change with huge profit.

Reply:
I don't think we should add big refactorings in `3.x`, that would delay the release.

Reply:
Big? It is 1h work.